satpy/readers/modis_l3.py

```diff
@@ -0,0 +1,132 @@
+
+There are multiple level 3 products, including some on sinusoidal grids and some on the climate modeling grid (CMG).
+This reader supports the CMG products at present, and the sinusoidal products will be added if there is demand.
+
+The reader has been tested with:
+    - MCD43c*: BRDF/Albedo data, such as parameters, albedo and nbar
+    - MOD09CMG: Surface Reflectance on climate monitoring grid.
+
+To get a list of the available datasets for a given file refer to the "Load data" section in :doc:`../reading`.
+
+"""
+import logging
+
+from pyresample import geometry
+
+from satpy.readers.hdfeos_base import HDFEOSGeoReader
+
+logger = logging.getLogger(__name__)
+
+
+class ModisL3GriddedHDFFileHandler(HDFEOSGeoReader):
+    """File handler for MODIS HDF-EOS Level 3 CMG gridded files."""
+
+    def _get_res(self):
+        """Compute the resolution from the file metadata."""
+        gridname = self.metadata["GridStructure"]["GRID_1"]["GridName"]
+        if "CMG" not in gridname:
+            raise ValueError("Only CMG grids are supported")
+
+        # Get the grid resolution from the grid name
+        pos = gridname.rfind("_") + 1
+        pos2 = gridname.rfind("Deg")
+
+        # Some products don't have resolution listed.
+        if pos < 0 or pos2 < 0:
+            self.resolution = 360. / self.ncols
+        else:
+            self.resolution = float(gridname[pos:pos2])
+
+    def _sort_grid(self):
+        """Get the grid properties."""
+        # First, get the grid resolution
+        self._get_res()
+
+        # Now compute the data extent
+        upperleft = self.metadata["GridStructure"]["GRID_1"]["UpperLeftPointMtrs"]
+        lowerright = self.metadata["GridStructure"]["GRID_1"]["LowerRightMtrs"]
+
+        # For some reason, a few of the CMG products multiply their
+        # decimal degree extents by one million. This fixes it.
+        if lowerright[0] > 1e6:
+            upperleft = tuple(val / 1e6 for val in upperleft)
+            lowerright = tuple(val / 1e6 for val in lowerright)
+
+        return (upperleft[0], lowerright[1], lowerright[0], upperleft[1])
+
+
+    def __init__(self, filename, filename_info, filetype_info, **kwargs):
+        """Init the file handler."""
+        super().__init__(filename, filename_info, filetype_info, **kwargs)
+
+        # Initialise number of rows and columns
+        self.nrows = self.metadata["GridStructure"]["GRID_1"]["YDim"]
+        self.ncols = self.metadata["GridStructure"]["GRID_1"]["XDim"]
+
+        # Get the grid name and other projection info
+        self.area_extent = self._sort_grid()
+
+
+    def available_datasets(self, configured_datasets=None):
+        """Automatically determine datasets provided by this file."""
+        logger.debug("Available_datasets begin...")
+
+        ds_dict = self.sd.datasets()
+
+        yield from super().available_datasets(configured_datasets)
+        common = {"file_type": "mcd43_cmg_hdf", "resolution": self.resolution}
```

Comment:
I'm not sure the debug log message is providing any useful information to the user or a log file. I also reordered when the `self.sd_datasets` is accessed since it isn't used until after the `yield from`.

```suggestion
        yield from super().available_datasets(configured_datasets)
        common = {"file_type": "mcd43_cmg_hdf", "resolution": self.resolution}
        ds_dict = self.sd.datasets()
```

Reply:
Changed to this (will refactor more once I sort out this dynamic datasets stuff.